06x-new-kafka-roller.md

```diff
@@ -0,0 +1,350 @@
+
+1. Initialise a context object for each node with the following data:
+```
+Context: {
+         nodeRef: <NodeRef object passed from KafkaReconciler>
+         nodeRoles: <This will be set using the pod labels `strimzi.io/controller-role` and `strimzi.io/broker-role>
+         state: UNKNOWN
+         lastTransition: <SYSTEM_TIME>
+         reason: <Result of the predicate function passed from KafkaReconciler>
+         numRestartAttempts: 0
+         numReconfigAttempts: 0
+         numRetries: 0
+}
+```
+Contexts are recreated in each reconciliation with the initial data above.
+
+2. Observe and transition each node's state to the corresponding state based on the information collected from the abstracted sources. If it can't retrieve any information from the sources, the reconciliation fails and the next reconciliation would start from step 1.
+
+3. If there are nodes in `NOT_READY` state, wait for them to have `SERVING` within the `postOperationTimeoutMs`.  
+   We want to give nodes chance to get ready before we try to connect to them or consider them for rolling. This is important especially for nodes which were just started.
+   This is consistent with how the current roller handles unready nodes.
+   - If the timeout is reached, proceed to the next step and check if any of the nodes need to be restarted.
+
+4. Group the nodes into the following categories based on their state and connectivity:
+   - `RESTART_FIRST` - Nodes that have `NOT_READY` or `NOT_RUNNING` state in their contexts. The nodes that we cannot connect to via Admin API will also be put into this group with its reason updated with `POD_UNRESPONSIVE`.
+   - `WAIT_FOR_LOG_RECOVERY` -  Nodes that have `RECOVERING` state.
+   - `RESTART` - Nodes that have non-empty list of reasons from the predicate function and have not been restarted yet (Context.numRestartAttempts == 0).
+   - `MAYBE_RECONFIGURE` - Broker nodes (including mixed nodes) that have an empty list of reasons and not been reconfigured yet (Context.numReconfigAttempts == 0).
+   - `NOP` - Nodes that have at least one restart or reconfiguration attempt (Context.numRestartAttempts > 0 || Context.numReconfigAttempts > 0 ) and have either
+             `LEADING_ALL_PREFERRED` or `SERVING` state.
+
+5. Wait for nodes in `WAIT_FOR_LOG_RECOVERY` group to finish performing log recovery. 
+   - Wait for nodes to have `SERVING` within the `postOperationTimeoutMs`. 
+   - If the timeout is reached for a node and its `numRetries` is greater than or equal to `maxRetries`, throw `UnrestartableNodesException` with the log recovery progress (number of remaining logs and segments). Otherwise increment node's `numRetries` and repeat from step 2.
+
+6. Restart nodes in `RESTART_FIRST` category:
+   - if one or more nodes have `NOT_RUNNING` state, we first need to check 2 special conditions:
+      - If all of the nodes are mixed and are in `NOT_RUNNING` state, restart them in parallel to give the best chance of forming the quorum.
```

Comment:
> Don't NOT_RUNNING nodes with dedicated controller node need parallel restart as well?
Initially, was only trying to solve the issue for the combined nodes because regular controllers don't have this issue (they have a different readiness check) but actually that's a good point. It does make sense to do the parallel restarts for the dedicated controller nodes too. That would simply cases such as cluster having dedicated and combined nodes, all of which are in pending state. 

> Should there be some sorting based on role to handle controllers first
Yes, we do sort them into groups first to maintain the order of controllers, mixed nodes and brokers.  I will make this clearer. 

> Can't all the NOT_RUNNING nodes be handled in parallel in general? They are not running, so there is no need to handle them one by one in any situation. Or?
That could be more efficient way to recover the NOT_RUNNING nodes that have old revisions.

Reply:
So I have tested the following changes:
- If all controllers are in NOT_RUNNING state, including pure and combined controllers, then we restart them in parallel. 
- All NOT_RUNNING nodes that have old revision will be handled in parallel as suggested but this includes nodes with any role, e.g. controller and broker nodes can get restarted at the same time. Do you think we should separate them by roles?
- When restarting nodes unresponsive to connections, in case there are more than one, they will be restarted one by one, in the order of pure controller, combined and broker. 

Do you think this changes make sense? If yes, I will go ahead and update the proposal.